agent/taskresource/envFiles/envfile.go

```diff
@@ -0,0 +1,432 @@
+	s3Client, err := envfile.s3ClientCreator.NewS3ClientForBucket(bucket, envfile.region, executionCredentials.GetIAMRoleCredentials())
+	if err != nil {
+		return errors.Wrapf(err, "unable to initialize s3 client for bucket %s", bucket)
+	}
+
+	seelog.Debugf("Downlading envfile with bucket name %v and key name %v", bucket, key)
+	// we save envfiles to path: /var/lib/ecs/data/envfiles/cluster_name/task_id/${s3filename.env}
+	downloadPath := filepath.Join(envfile.resourceDir, key)
+	err = envfile.writeEnvFile(func(file oswrapper.File) error {
+		return s3.DownloadFile(bucket, key, s3DownloadTimeout, file, s3Client)
+	}, downloadPath)
+
+	if err != nil {
+		return errors.Wrapf(err, "unable to download env file with key %s from bucket %s", key, bucket)
+	}
+
+	seelog.Debugf("Downloaded envfile from s3 and saved to %s", downloadPath)
+	return nil
+}
+
+func (envfile *EnvironmentFileResource) writeEnvFile(writeFunc func(file oswrapper.File) error, fullPathName string) error {
+	// File moves (renaming) are atomic while file writes are not
+	// so we write to a temp file before renaming to actual file
+	// multiple programs calling TempFile will not reference the same file
+	// so this should be ok to be called by multiple go routines
+	tmpFile, err := envfile.ioutil.TempFile(envfile.resourceDir, envTempFilePrefix)
+	if err != nil {
+		seelog.Infof("something went wrong trying to create a temp file with prefix %s", envTempFilePrefix)
+		return err
+	}
+	defer tmpFile.Close()
+
+	err = writeFunc(tmpFile)
+	if err != nil {
+		seelog.Infof("something went wrong trying to write to tmpFile %s", tmpFile.Name())
+		return err
+	}
+
+	// persist file to disk
+	err = tmpFile.Sync()
+	if err != nil {
+		seelog.Infof("something went wrong trying to persist envfile to disk")
+		return err
+	}
+
+	err = envfile.os.Rename(tmpFile.Name(), fullPathName)
+	if err != nil {
+		seelog.Infof("something went wrong when trying to rename envfile from %s to %s", tmpFile.Name(), fullPathName)
+		return err
+	}
+
+	return nil
+}
+
+func (envfile *EnvironmentFileResource) getEnvironmentFilesSource() []apicontainer.EnvironmentFile {
+	envfile.lock.RLock()
+	defer envfile.lock.RUnlock()
+
+	return envfile.environmentFilesSourceUnsafe
+}
+
+func (envfile *EnvironmentFileResource) getExecutionCredentialsID() string {
+	envfile.lock.RLock()
+	defer envfile.lock.RUnlock()
+
+	return envfile.executionCredentialsID
+}
+
+// Cleanup removes env files downloaded for the task
+func (envfile *EnvironmentFileResource) Cleanup() error {
+	// TODO
+	return nil
+}
+
+type EnvironmentFileResourceJSON struct {
+	TaskARN                string                         `json:"taskARN"`
+	CreatedAt              *time.Time                     `json:"createdAt,omitempty"`
+	DesiredStatus          *EnvironmentFileStatus         `json:"desiredStatus"`
+	KnownStatus            *EnvironmentFileStatus         `json:"knownStatus"`
+	EnvironmentFilesSource []apicontainer.EnvironmentFile `json:"envfileResources"`
```

Comment:
nit: maybe envFileSources? should match the field name

Reply:
made sure matches